### ISSUE TITLE
Access Settings from a Project plugin (e.g., Project.getSettings())

### ISSUE THREAD

**Issue by anuraaga** [gradle/gradle#17295]
### Expected Behavior

Ability to access `Settings` via `Project`

### Current Behavior

There doesn't seem to be any way. An attempted workaround hoping `afterSettings` is executed inline also doesn't work (is  that itself a bug?).

Doesn't work:
```java
AtomicReference<Settings> settings = new AtomicReference<>();
project.getGradle().afterSettings(settings::set);
```

### Context

I have a plugin that needs to know the repositories set for a project. After moving repositories from the project (one of many in a monorepo) to `dependencyResolutionManagement`, this stopped working since `Project.getRepositories()` does not include them. This would be fine if I could access that block's information in the plugin but I can't find a way. Since `Settings` is always evaluated before projects, I would expect there to be a `Project.getSettings()`.

**Comment by vlsi** [comment]
Just in case, the issue affects `refreshVersions` (cc @jmfayard, @LouisCAD  ) plugin as well: https://github.com/jmfayard/refreshVersions/blob/65d564f60dcda50e3dda5843e5b5aa134e425fea/plugins/core/src/main/kotlin/de/fayard/refreshVersions/core/internal/VersionsPlaceholdersReplacement.kt#L183

The error there is "empty fetchers list" yet the true meaning is "no MavenRepositories found"

```
        java.lang.IllegalArgumentException: Cannot get version candidates with an empty fetchers list.
            at de.fayard.refreshVersions.core.internal.GettingVersionCandidatesKt.getVersionCandidates(GettingVersionCandidates.kt:63)
            at de.fayard.refreshVersions.core.internal.GettingVersionCandidatesKt.getVersionCandidates(GettingVersionCandidates.kt:16)
            at de.fayard.refreshVersions.core.internal.VersionsPlaceholdersReplacementKt$Write versions candidates using latest most stable version and get it$2.invokeSuspend(VersionsPlaceholdersReplacement.kt:198)
```

**Comment by LouisCAD** [comment]
@vlsi No open issue for that?? 🤔

**Comment by anuraaga** [comment]
@ben-manes gradle dependencies plugin gets hit too. Perhaps in addition or alternative to this we need a `getResolvedRepositories` as it seems to be used by many plugins.

**Comment by blogcommoda** [comment]
,z.

Gilson

Em sex, 28 de mai de 2021 11:12, Louis CAD ***@***.***>
escreveu:

**Comment by ben-manes** [comment]
@anuraaga We're not affected by this except for an incorrect log message, where we print the repositories for debugging purposes. The plugin avoids mucking with them otherwise due to previous bad experiences with bugs and api limitations. Instead we copy the configuration to resolve with so that these types of hidden details work automagically (assuming copy() isn't broken, which it often is).

**Comment by anuraaga** [comment]
@ben-manes Sorry, yeah I had forgotten that while I had a problem first time I ran upgrading solved it nicely.

**Comment by TWiStErRob** [comment]
Google's Android Gradle Plugin is also affected ([commit that added](https://android.googlesource.com/platform/tools/base/+/40810519ad185faa1d1ef44db3f2db81b79b37b7%5E%21/build-system/gradle-core/src/main/java/com/android/build/gradle/internal/tasks/PerModuleReportDependenciesTask.kt), [line](https://android.googlesource.com/platform/tools/base/+blame/studio-main/build-system/gradle-core/src/main/java/com/android/build/gradle/internal/tasks/PerModuleReportDependenciesTask.kt#311)).

**Comment by TWiStErRob** [comment]
@grossws this does not execute in `Plugin.apply(Project)`, it's too late the settings is already evaluated (raised https://github.com/gradle/gradle/issues/20047):
```
project.gradle.settingsEvaluated { settings ->
	println(settings.dependencyResolutionManagement.repositoriesMode.get())
}
```

Building on AGP 7.x's "solution" I came up with this hack:
```
import org.gradle.api.Project
import org.gradle.api.initialization.Settings
import org.gradle.api.internal.GradleInternal

val Project.settings: Settings
	get() = (gradle as GradleInternal).settings
```
This way I can access `project.settings.dependencyResolutionManagement.repositoriesMode.get()` at any time, and it works all fine, but would love to see this issue to expose this properly go through, maybe even a subset of the interface (e.g. without mutability).

**Comment by liutikas** [comment]
Note, this is blocking our work on Gradle project isolation, as we are unable to move from using `allprojects { repositories { ... ` to settings.gradle definition of repositories.

**Comment by liutikas** [comment]
To work around it for us we went with `settings.gradle`

```
getGradle().beforeProject {
  it.repositories { ... }
}
```

while this is not as good as using `dependencyResolutionManagement.repositories` at least it unblocks project isolation work.

**Comment by alexstaeding** [comment]
I am also stumbling across this strange limitation in the API. My use case is serializing the dependency configuration along with the repository configuration so that it is shipped with an artifact to be re-exported back into a full Gradle project.

Currently, it seems the only way to do this (using @TWiStErRob's suggestion) is something along the lines of:
```kotlin
    @get:Input
    @Suppress("UnstableApiUsage")
    val repositories: ListProperty<Pair<String, String>> = project.objects.listProperty<Pair<String, String>>().value(
        (project.gradle as GradleInternal).settings.dependencyResolutionManagement.repositories.mapToPairs() +
            project.repositories.mapToPairs()
    )

    private fun RepositoryHandler.mapToPairs(): List<Pair<String, String>> =
        filterIsInstance<MavenArtifactRepository>().map { it.name to it.url.toString() }
```

See https://github.com/sourcegrade/jagr/pull/120#issuecomment-1264279939

**Comment by staktrace** [comment]
I'm also running into this limitation of not being able to access the settings object from a project plugin. In my case I would like to know if the remote build cache is enabled so I can tag the buildscan accordingly. I know I could use a settings plugin to do this but that's not practical for my particular situation as it would involve rolling it out to hundreds of repos.

**Comment by liutikas** [comment]
Any updates here? Thus bug forces folks to use worse APIs.

**Comment by ljacomet** [comment]
Sorry for the late reply!

This issue will be closed as a duplicate of

* #16616 

Please add your use case and 👍 to that issue.

***

Closing this as a duplicate because the use case is really focused on repositories.

If you believe the settings access is needed for other use cases, or that direct access to settings repositories is needed from a project, please open a more specific issue for this, possibly relating it to this one.
